src/extension-runners/firefox-desktop.js

```diff
@@ -0,0 +1,229 @@
+import type FirefoxProfile from 'firefox-profile';
+
+import {
+  MultipleExtensionsReloadError,
+  RemoteTempInstallNotSupported,
+  WebExtError,
+} from '../errors';
+import * as defaultFirefoxApp from '../firefox';
+import {
+  connectWithMaxRetries as defaultFirefoxConnector,
+} from '../firefox/remote';
+import {createLogger} from '../util/logger';
+// Import flow types from project files.
+import type {
+  FirefoxRDPResponseAddon,
+  RemoteFirefox,
+} from '../firefox/remote';
+import type {
+  ExtensionRunnerParams,
+} from './base';
+import type {
+  FirefoxPreferences,
+} from '../firefox/preferences';
+import type {
+  FirefoxInfo, // eslint-disable-line import/named
+} from '../firefox/index';
+
+export type FirefoxDesktopExtensionRunnerParams = ExtensionRunnerParams & {
+  // Firefox desktop CLI params.
+  customPrefs?: FirefoxPreferences,
+  browserConsole: boolean,
+  firefoxBinary: string,
+  preInstall: boolean,
+
+  // Firefox desktop injected dependencies.
+  firefoxApp: typeof defaultFirefoxApp,
+  firefoxClient: typeof defaultFirefoxConnector,
+};
+
+const log = createLogger(__filename);
+
+/**
+ * This module provide an ExtensionRunner subclass that manage an extension executed
+ * in a Firefox for Desktop instance.
+ */
+
+export class FirefoxDesktopExtensionRunner {
+  params: FirefoxDesktopExtensionRunnerParams;
+  reloadableExtensions: Map<string, string>;
+  cleanupCallbacks: Set<Function>;
+  profile: FirefoxProfile;
+  runningInfo: FirefoxInfo;
+  remoteFirefox: RemoteFirefox;
+
+  constructor(
+    params: FirefoxDesktopExtensionRunnerParams,
+  ) {
+    this.params = params;
+
+    this.reloadableExtensions = new Map();
+    this.cleanupCallbacks = new Set();
+  }
+
+  // Method exported from the IExtensionRunner interface.
+
+  async run(): Promise<void> {
+    // Get a firefox profile with the custom Prefs set (a new or a cloned one).
+    // Pre-install extensions as proxy if needed (and disable auto-reload if you do)
+    await this.setupProfileDir();
+
+    // (if reload is enabled):
+    // - Connect to the firefox instance on RDP
+    // - Install any extension if needed (if not installed as proxy)
+    // - Keep track of the extension id assigned in a map with the sourceDir as a key
+    await this.startFirefoxInstance();
+  }
+
+  async reloadAllExtensions(): Promise<void> {
+    const reloadErrors = new Map();
+    for (const {sourceDir} of this.params.extensions) {
+      try {
+        await this.reloadExtensionBySourceDir(sourceDir);
+      } catch (error) {
+        reloadErrors.set(sourceDir, error);
+      }
+    }
+
+    if (reloadErrors.size > 0) {
+      return Promise.reject(new MultipleExtensionsReloadError(reloadErrors));
+    }
+  }
+
+  async reloadExtensionBySourceDir(extensionSourceDir: string): Promise<void> {
+    const addonId = this.reloadableExtensions.get(extensionSourceDir);
+    if (!addonId) {
+      throw new WebExtError('Extension not reloadable');
```

Comment:
I think it would be better to make this message more explicit, like:

```js
`No addonId has been mapped to extension at source dir "${extensionSourceDir}"`
```